feature: rstuf-worker for managing targets via SQL (#143) (#145)
vmware/repository-service-tuf-worker: docs/source/devel/index.rst

- **Bump Roles** that contain online keys ("Snapshot", "Timestamp" and Hahsed
  Bins ("bins-XX").
- **Publish the new Hashed Bins Target Roles** ("bins-XX") with new/removed
  targets.

About **Bump Roles** (``bump_online_roles``) that contain online keys is easy.
These roles have short expiration (defined during repository configuration) and
must be "bumped" frequently. The implementation in the RepositoryMetadata

**Publish the new Hashed Bins Target Roles** (``publish_targets``) is part of the
solution for the :ref:`Repository Worker scalability, Issue 17
<devel/known_issues:(Solved) Scalability>`.

To understand more, every time the API sends a task to add a new target, the
Hashed Bins Roles must be changed to add the new target(s), followed by a new

> Snapshot and Timestamp versions.

Reviewer: Even though timestamp, snapshot and hashed bins are important for us I don't see why they are with upper case letters.
I have the same opinions for `Repository Worker` and `Repository Metadata backend`.

What do you think?

Author: I use it to highlight the role name (not to confuse it with the generic word snapshot timestamp) and the component name (Repository Worker).
We do it across many docs. My suggestion is if we want to normalize it do a specific issue and change across all doc.

Reviewer: Okay, let's leave it like that for now.

   and `remove_targets
   <repository_service_tuf_worker.html#repository_service_tuf_worker.repository.MetadataRepository.remove_targets>`_

Repository Worker adds/removes the target to the SQL Database.

It means the multiple Repository Workers can write multiple Targets
(``TargetFiles``) simultaneously from various tasks in the Database.

When a task finishes, it send a task the ``publish_targets``.

Every minute, the routine task **Publish the new Hashed Bins Roles** also runs.

The task will continue run, wait until all the targets are persisted to the
Repository Metadata backend.

The **Publish the new Hashed Bins Roles** task (``publish_targets``) runs once

> per time to using locks [#f1]_ . It will  will do:

Reviewer: Could you add a sentence or two about how we synchronize the state of the database with the state of the JSON metadata files?

Author: @MVrachev  Added the mention.

   and `remove_targets
   <repository_service_tuf_worker.html#repository_service_tuf_worker.repository.MetadataRepository.remove_targets>`_

Repository Worker adds/removes the target to the SQL Database.

It means the multiple Repository Workers can write multiple Targets
(``TargetFiles``) simultaneously from various tasks in the Database.

The **Publish the new Hashed Bins Target Roles** is a synchronization between
the SQL Database and the Hashed Bins Target Roles in the Backend Storage (i.e.
JSON files in the filestytem)

> When a task finishes, it send a task the **Publish the new Hashed Bins Target

Reviewer: ```suggestion
When a task finishes, it sends a task the **Publish the new Hashed Bins Target
```

Roles** .

Every minute, the routine task **Publish the new Hashed Bins Target Roles**
also runs.

> The task will continue run, wait until all the targets are persisted to the

Reviewer: ```suggestion
The task will continue running and waiting until all the targets are persisted to the
```

Repository Metadata backend.

> The **Publish the new Hashed Bins Target Roles** runs once per time to using

Reviewer: ```suggestion
The **Publish the new Hashed Bins Target Roles** task runs once per time to using
```
